components/supervisor/pkg/supervisor/tasks.go

```diff
@@ -0,0 +1,412 @@
+			continue
+		}
+
+		taskLog.Info("task terminal has been started")
+		tm.updateState(func() *task {
+			t.Terminal = resp.Alias
+			t.State = api.TaskState_running
+			return t
+		})
+
+		go func(t *task) {
+			terminal.Command.Process.Wait()
+			taskLog.Info("task terminal has been closed")
+			tm.setTaskState(t, api.TaskState_closed)
+		}(t)
+
+		if runContext.headless {
+			tm.watch(t, terminal)
+		}
+		terminal.PTY.Write([]byte(t.command + "\r\n"))
+	}
+
+	if runContext.headless {
+		tm.report(ctx)
+	}
+}
+
+func (task *task) getCommand(context *runContext) string {
+	commands := task.getCommands(context)
+	command := composeCommand(composeCommandOptions{
+		commands: commands,
+		format:   "{\r\n%s\r\n}",
+		sep:      " && ",
+	})
+
+	if strings.TrimSpace(command) == "" {
+		return ""
+	}
+
+	if context.headless {
+		// it's important that prebuild tasks exit eventually
+		// also, we need to save the log output in the workspace
+		return command + "; exit"
+	}
+
+	histfile := "/workspace/.gitpod/cmd-" + task.Id
+	histfileCommands := commands
+	if context.contentSource == csapi.WorkspaceInitFromPrebuild {
+		histfileCommands = []*string{task.config.Before, task.config.Init, task.config.Prebuild, task.config.Command}
+	}
+	err := ioutil.WriteFile(histfile, []byte(composeCommand(composeCommandOptions{
+		commands: histfileCommands,
+		format:   "%s\r\n",
+	})), 0644)
+	if err != nil {
+		log.WithField("histfile", histfile).WithError(err).Fatal("cannot write histfile")
+		return command
+	}
+	// the space at beginning of the HISTFILE command prevents the HISTFILE command itself from appearing in
+	// the bash history.
+	return " HISTFILE=" + histfile + " history -r; " + command
+}
+
+func (task *task) getCommands(context *runContext) []*string {
+	if context.headless {
+		// prebuild
+		return []*string{task.config.Before, task.config.Init, task.config.Prebuild}
+	}
+	if context.contentSource == csapi.WorkspaceInitFromPrebuild {
+		// prebuilt
+		prebuildLogFileName := task.prebuildLogFileName()
+		legacyPrebuildLogFileName := "/workspace/.prebuild-log-" + task.Id
+		printlogs := "[ -r " + legacyPrebuildLogFileName + " ] && cat " + legacyPrebuildLogFileName + "; [ -r " + prebuildLogFileName + " ] && cat " + prebuildLogFileName + "; true"
+		return []*string{task.config.Before, &printlogs, task.config.Command}
+	}
+	if context.contentSource == csapi.WorkspaceInitFromBackup {
+		// restart
+		return []*string{task.config.Before, task.config.Command}
+	}
+	// init
+	return []*string{task.config.Before, task.config.Init, task.config.Command}
+
+}
+
+func (task *task) prebuildLogFileName() string {
+	return "/workspace/.gitpod/prebuild-log-" + task.Id
+}
+
+func (tm *tasksManager) watch(task *task, terminal *terminal.Term) {
+	workspaceLog := log.WithField("component", "workspace")
+	stdout := terminal.Stdout.Listen()
+	start := time.Now()
+	prebuildChan := make(chan bool)
+	task.prebuildChan = &prebuildChan
+	go func() {
```

Comment:
If I understand correctly, at the end of this go routine you must have written to `prebuildChan`, correct?
In that case, you could add a `success` variable and do something like
```go
defer func() {
    prebuildChan <- success
}()
```
setting `success` to the appropriate value. This would the contract/intention clearer that `prebuildChan` must be written in this function.